### PR TITLE
fixes issue #2, calling super async when checking jobs over recursing

### PR DESCRIPTION
**Pull request by abrandoned**
@quixoten @brianstien @film42

### CODE REVIEW

File: lib/lifeguard/infinite_threadpool.rb [hunk 1]
```diff
@@ -28,7 +32,7 @@ def check_queued_jobs
       return if @queued_jobs.size <= 0
 
       queued_job = @queued_jobs.pop
-      async(*queued_job[:args], &queued_job[:block])
+      super_async(*queued_job[:args], &queued_job[:block])
       check_queued_jobs
```
Comment (film42) [inline lib/lifeguard/infinite_threadpool.rb]: Let's turn this recursive call into a loop.

**Comment by quixoten** [comment]
Is there a potential race condition where two threads could get to `super_async` at the same time? If one of those jobs gets the last remaining thread, the other would be thrown away.

File: lib/lifeguard/infinite_threadpool.rb [hunk 2]
```diff
@@ -27,8 +32,12 @@ def check_queued_jobs
...
 
-      queued_job = @queued_jobs.pop
-      async(*queued_job[:args], &queued_job[:block])
+      @super_async_mutex.synchronize do
+        return if busy?
+        queued_job = @queued_jobs.pop
+        super_async(*queued_job[:args], &queued_job[:block])
+      end
+
       check_queued_jobs
```
Comment (film42) [inline lib/lifeguard/infinite_threadpool.rb]: I still think this recursive call is a problem. A loop would ensure we never blow the stack.

Comment (film42) [reply lib/lifeguard/infinite_threadpool.rb]: Oh hey, write as I comment. Nice!

**Comment by quixoten** [comment]
I think there's still an issue, because [this](https://github.com/mxenabled/lifeguard/blob/cf40be5d8557aa4fb003cf72b3d947dc12649c48/lib/lifeguard/infinite_threadpool.rb#L21) has the potential to return immediately and beat the `super_synchronized` call to `super_async`.

**Comment by quixoten** [comment]
:shipit:

**Pull request merged**